packages/xstate-wallet/src/channel-wallet.ts

```diff
@@ -235,134 +224,60 @@ export class ChannelWallet {
-      channelFunding.status !== 'DEPOSITED'
-    ) {
-      response.deposit = true;
-      return response;
-    }
-
-    // Should we sign the postfund state?
-    if (
-      BN.gte(channelFunding.amountOnChain, depositInfo.fundedAt) &&
-      latestState.turnNum === 3 &&
-      channel.latestSignedByMe.turnNum === 0
-    ) {
-      const newState = channel.signAndAdd(latestState, pk);
-      response.signedStates = [newState];
-      return response;
-    }
-
-    return response;
-  }
+  /**
+   *  START of wallet 2.0
+   */
 
-  async onFundingUpdate(channelId: string, channelChainInfo: ChannelChainInfo): Promise<void> {
-    this.channelFunding[channelId] = {
-      ...this.channelFunding[channelId],
-      amountOnChain: channelChainInfo.amount
-    };
-    await this.onOpenChannelObjective(channelId);
-  }
+  private async crankRichObjectives(event: DirectFunder.OpenChannelEvent): Promise<void> {
+    const richObjectives = this.store.richObjectives;
+    for (const channelId of Object.keys(richObjectives)) {
+      const richObjective = richObjectives[channelId];
+      const result = DirectFunder.openChannelCranker(
+        richObjective,
+        event,
+        await this.store.getPrivateKey(await this.store.getAddress())
+      );
 
-  async getDepositInfo(channelId: string): Promise<DepositInfo> {
-    const {latestState, myIndex} = await this.store.getEntry(channelId);
-    const {allocationItems} = checkThat(latestState.outcome, isSimpleEthAllocation);
-
-    const fundedAt = allocationItems.map(a => a.amount).reduce(BN.add);
-    let depositAt = Zero;
-    for (let i = 0; i < allocationItems.length; i++) {
-      const {amount} = allocationItems[i];
-      if (i !== myIndex) depositAt = BN.add(depositAt, amount);
-      else {
-        const totalAfterDeposit = BN.add(depositAt, amount);
-        return {depositAt, myDeposit: amount, totalAfterDeposit, fundedAt};
+      richObjectives[channelId] = result.objective;
+
+      for (const action of result.actions) {
+        switch (action.type) {
+          case 'sendStates':
+            await Promise.all(action.states.map(state => this.store.addState(state, true)));
+            break;
+          case 'deposit':
+            if (this.store.depositsSubmitted[channelId]) {
+              throw new Error(
+                `Attempting to submit a deposit for a channel with already submitted deposit ${this.store.depositsSubmitted[channelId]}`
+              );
+            }
+            const fundingMilestones = DirectFunder.utils.fundingMilestone(
+              richObjective.openingState,
+              richObjective.openingState.participants[richObjective.myIndex].destination
+            );
+
+            // Record that a deposit will be made
+            this.store.depositsSubmitted[channelId] = {
+              amountOnChain: fundingMilestones.targetBefore,
+              amountDeposited: action.amount
+            };
```

Comment:
Do you think this information should be recorded in the rich objective itself?

Reply:
That's not a bad idea! 

On a somewhat related thought, I wonder if the rich objective data structure should be split up into shareable objective data (data that applies to all participants) and participant specific data. This split would make it easier to send around rich objectives between participants if we chose to do so in the future.

Perhaps a premature optimization?

Reply:
Filed https://github.com/statechannels/statechannels/issues/3469.

Reply:
> Perhaps a premature optimization?

I think you're definitely on the right track, but it feels like a premature optimization.

There's also an argument to sharing _your deposit transaction_ with your peers in the channel, as that enables their chain service to monitor that particular transaction.